### ISSUE TITLE
Undocumented change to Source RCON Protocol?

### ISSUE THREAD

**Issue by smhxx** [koraktor/steam-condenser#169]
I'm not sure when this change was made, but it appears that there's been a small undocumented change to the Source RCON protocol. It doesn't break anything (probably why nobody caught it,) but I thought I'd let you know since it appears to cause a few unreachable lines of code in steam-condenser, and probably some unintended functionality. I just stumbled across it a couple hours ago while I was sniffing through the packets my server was sending.

Basically, SRCDS will no longer send a `SERVERDATA_AUTH_RESPONSE` in response to a `SERVERDATA_EXECCOMMAND`, even if the connection has not been authenticated, or if the last auth request was unsuccessful. Instead, the server now sends an empty `SERVERDATA_RESPONSE_VALUE`. It also appears that SRCDS will no longer ban users for **any** `EXECCOMMAND` packet; I bombarded a server on my local machine with several hundred requests for "sv_cheats" at a 50ms interval, and it didn't ban me until I sent an auth request afterwards. Combined with the fact that SRCDS will drop all RCON connections when `rcon_password` changes, I don't think it's possible at all to get an `AUTH_RESPONSE` in reply to an `EXECCOMMAND`. Even if the user's IP is manually added to the banlist while they are connected, SRCDS will continue to accept and reply to RCON commands until the connection is closed, at which point it will refuse any attempt to open a new connection (as it normally would for IP banned users.)

Here's a quick snippet to show you what I'm talking about. It's written in Node but I think you can see what's going on here. The first time I send the command, before authing, it replies with an empty `SERVERDATA_RESPONSE_VALUE`... then I auth, and it responds with the expected cvar description. Everything that passes through the socket during this test is printed to stdout; I temporarily disabled terminators just for the sake of clarity.

```
> rcon.connect("74.91.115.129:27015");
undefined

> rcon.sendCommand("sv_cheats")
undefined

Sending: <Buffer 13 00 00 00 00 00 00 00 02 00 00 00 73 76 5f 63 68 65 61 74 73 00 00>
         { type: 2, id: 0, body: 'sv_cheats', callback: undefined }

Received: <Buffer 0a 00 00 00 00 00 00 00 00 00 00 00 00 00>
          { type: 0, id: 0, body: '' }

> rcon.sendPassword("correctpassword")
undefined

Sending: <Buffer xx 00 00 00 01 00 00 00 03 00 00 00 xxxxxxxxxx 00 00>
         { type: 3, id: 1, body: 'correctpassword', callback: undefined }

Received: <Buffer 0a 00 00 00 01 00 00 00 00 00 00 00 00 00>
          { type: 0, id: 1, body: '' }

Received: <Buffer 0a 00 00 00 01 00 00 00 02 00 00 00 00 00>
          { type: 2, id: 1, body: '' }

> rcon.sendCommand("sv_cheats")
undefined

Sending: <Buffer 13 00 00 00 02 00 00 00 02 00 00 00 73 76 5f 63 68 65 61 74 73 00 00>
         { type: 2, id: 2, body: 'sv_cheats', callback: undefined }

Received: <Buffer 49 00 00 00 02 00 00 00 00 00 00 00 22 73 76 5f 63 68 65 61 74 73 22 20 3d 20 22 30 22 0a 20 6e 6f 74 69 66 79 20 72 65 70 6c 69 63 61 74 65 64 0a 20 2d ...>
          { type: 0,
            id: 2,
            body: '"sv_cheats" = "0"\n notify replicated\n - Allow cheats on server\n' }
```

I'm not sure when this change happened, or if the documentation was just plain wrong from the beginning, but I wouldn't be the least bit surprised if it were made years ago and nobody ever noticed. The wiki page on developer.valvesoftware.com was apparently copy/pasted from some forum post made a bajillion years ago and isn't maintained by anyone from Valve. Anyway, check it out yourself if you want to - make sure I'm not crazy or just imagining things.

-- Speedhaxx

**Comment by smhxx** [comment]
I guess this oddity is also somewhat related, in a way: attempting to change the RCON password remotely results in an `RCONBanError` being thrown, even though the command is executed successfully and the user is not added to the banlist. Perhaps it misinterprets the connection being dropped when the RCON password changes as the user being IP banned?

```
irb(main):005:0> server = SourceServer.new("127.0.1.1")
=> #<SourceServer:0xb7012590 @rcon_socket=#<RCONSocket:0xb701248c @ip="127.0.1.1", @port=27015>, @ip_address="127.0.1.1", @ip_index=0, @port=27015, @socket=#<SourceSocket:0xb7012478 @socket=#<UDPSocket:0xb701243c>>, @ip_addresses=["127.0.1.1"], @host_names=["PubCompVM"], @rcon_authenticated=false>
irb(main):006:0> server.rcon_auth("password")
=> true
irb(main):007:0> server.rcon_exec("rcon_password anotherpass")
RCONBanError: You have been banned from this server.
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/sockets/rcon_socket.rb:79:in `reply'
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/servers/source_server.rb:87:in `rcon_exec'
    from (irb):7
    from :0
irb(main):008:0> server = SourceServer.new("127.0.1.1")
=> #<SourceServer:0xb7009968 @rcon_socket=#<RCONSocket:0xb7009864 @ip="127.0.1.1", @port=27015>, @ip_address="127.0.1.1", @ip_index=0, @port=27015, @socket=#<SourceSocket:0xb7009850 @socket=#<UDPSocket:0xb7009814>>, @ip_addresses=["127.0.1.1"], @host_names=["PubCompVM"], @rcon_authenticated=false>
irb(main):009:0> server.rcon_auth("anotherpass")
=> true
```

**Comment by koraktor** [comment]
Thanks for the heads up.

The problem here, especially the effect when changing the RCON password, is that the protocol doesn't seem to allow an intelligent handling of those problems. I already made a somewhat big statement about those problems in issue #128.

Steam Condenser's current behavior is the best I could come up with, having a minimal trade-off for the user. Unauthenticated connections are handled internally, so you aren't able to do something like you did in your initial code above.

I you have any suggestions to further improve this, please let me know.

**Comment by smhxx** [comment]
The only thing that really jumped out at me is that since SRCDS now closes the connection if auth becomes invalid rather than sending `AUTH_RESPONSE`s until the client re-authenticates, the following bit of code in `/lib/steam/servers/source_server.rb` is unreachable/unneeded:

``` ruby
response_packet = @rcon_socket.reply
# If the connection's auth became invalid, @rcon_socket would have been closed by the server
# SRCDS won't send an AUTH_RESPONSE in reply to an EXECCOMMAND anymore, so this code is unreachable
if response_packet.is_a? RCONAuthResponse
  @rcon_authenticated = false
  raise RCONNoAuthError
end
```

Here's what actually happens if auth becomes invalid while steam-condenser is connected:

```
irb(main):001:0> require "rubygems"
=> true
irb(main):002:0> require "steam-condenser"
=> true
irb(main):003:0> server = SourceServer.new("127.0.1.1")
=> #<SourceServer:0xb70c85c0 @rcon_socket=#<RCONSocket:0xb70c8160 @ip="127.0.1.1", @port=27015>, @ip_address="127.0.1.1", @ip_index=0, @port=27015, @socket=#<SourceSocket:0xb70c814c @socket=#<UDPSocket:0xb70c8110>>, @ip_addresses=["127.0.1.1"], @host_names=["PubCompVM"], @rcon_authenticated=false>
irb(main):004:0> server.rcon_auth("password")
=> true
irb(main):005:0> server.rcon_exec("sv_cheats")
=> "\"sv_cheats\" = \"0\"\n notify replicated\n - Allow cheats on server"
irb(main):006:0> server.rcon_exec("status")
Errno::EPIPE: Broken pipe - send(2)
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/sockets/steam_socket.rb:83:in `send'
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/sockets/steam_socket.rb:83:in `send'
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/sockets/rcon_socket.rb:66:in `send'
    from /var/lib/gems/1.8/gems/steam-condenser-1.2.1/lib/steam/servers/source_server.rb:83:in `rcon_exec'
    from (irb):6
    from :0
```

**EDIT:** As for fixing it, the only "smart" solution I can think of would be:
- When `rcon_auth()` is called, remember the password provided by the user
- If the socket gets closed by the other end, try to re-open it once
- If the new connection is accepted, try to auth once using the previously provided password
- Since SRCDS closes the connection _any time_ `rcon_password` is written to, even if the password is the same, we may still have the correct password (i.e. if server.cfg was executed on mapchange)
- If it fails, throw an error saying the password has changed
- Otherwise, the connection has been re-established and is ready to send commands again

Or, if that's too hands-on a solution for you, simply throw an RCONPasswordChangedError or something if the socket gets closed by the other end.

**Comment by koraktor** [comment]
I know about that problem, in Ruby you'll get a `RCONBanError` like your second comment demonstrates.

How did you trigger the connection drop in your last sample? I think the `EPIPE` should not happen.

**Comment by smhxx** [comment]
I changed the RCON password from the server console.

**Comment by smhxx** [comment]
Oops, wrong button. xD

**Comment by smhxx** [comment]
I just verified that you also get the `EPIPE` if the map changes while you are connected (because server.cfg is executed on mapchange, which usually sets the rcon_password cvar).

**Comment by koraktor** [comment]
Hmm, seems like the TCP connection isn't completely reset, only the reading end (of the server):

```
irb(main):001:0> s = SourceServer.new '192.168.0.5'
=> #<SourceServer:0x1027af348 @ip_index=0, @rcon_socket=#<RCONSocket:0x1027af140 @ip="192.168.0.5",@port=27015>, @socket=#<SourceSocket:0x1027af118 @socket=#<UDPSocket:0x1027af0a0>>, @ip_addresses=["192.168.0.5"], @host_names=["192.168.0.5"], @rcon_authenticated=false, @ip_address="192.168.0.5", @port=27015>
irb(main):002:0> s.rcon_auth 'test'
=> true
irb(main):003:0> sock = s.instance_variable_get(:@rcon_socket).instance_variable_get(:@socket)
=> #<TCPSocket:0x1027a3458>
# Resetting the password now
irb(main):004:0> sock.recv 0
=> ""
irb(main):005:0> sock.send "", 0
Errno::EPIPE: Broken pipe - send(2)
        from (irb):5:in `send'
        from (irb):5
```

So in fact this could be handled easily and we could get an overall good behavior.

But to be honest, I don't want to workaround buggy behavior just to have Valve fix it in SrcDS in a few days (OK, it seems this problem exists since several weeks or months).

Therefore, I'm going to report this on the HLDS mailing list.

**Comment by smhxx** [comment]
Oh, so it just drops the connection immediately and leaves it half-open? I'd think it would at least go through the proper handshakes to end the connection, but your test shows otherwise. That does seem awfully strange; I'm not sure whether that behavior was intended or not. I'd definitely put this issue on hold until we get an answer from Valve on whether that behavior is actually correct, or if it's a bug that nobody's noticed until now. Like you said, there's no point in coding to the current behavior if it's just going to get patched once they realize what's going on. Let me know if you get an official response on the matter, I'm interested to know what the answer is for my own module as well.

~~**EDIT:** Just confirmed via `tcpdump` that SRCDS does _not_ send FIN when a connection is dropped due to the RCON password changing. Seems like it probably should, you would think...~~

[**"Scratch that, reverse it." -Willy Wonka**](https://github.com/koraktor/steam-condenser/issues/169#issuecomment-10046018)

**Comment by koraktor** [comment]
I just sent a somewhat lengthy mail to HLDS mailing list.
Usually, there's at least one Valve employee replying to serious problem reports. If not, I'll contact some of them directly.

Stay tuned. :clock11:

**Comment by smhxx** [comment]
It's hard to tell, but #145 bears a strange similarity to what we're experiencing here. The timeout error is obviously expected, since the server is in the middle of a map change, but the pipe breaks after the server comes back from the `changelevel`. Could that be the same `rcon_password` issue we're observing here? If so, this "bug" has existed in SRCDS for at least 9 months... that makes me wonder whether it's actually intentional after all. I'm not sure why they'd want it to just drop connections like that, but if it's been doing this for that long, perhaps there's a reason for it.

**Comment by koraktor** [comment]
Nice, I knew there where other related issues, but didn't find the time to check them, yet.

I think the connection reset in #144 was a "real" connection reset (i.e. `ECONNRESET`) and not the half closed connection we see now. That's at least consistent, but probably the best we can get for server shutdown / restart.
Map and password changes should not break RCON, though.

**Comment by smhxx** [comment]
Well, I went back and took a second look at the `tcpdump` and, as it turns out, I was dead wrong before. Here's what happens when the server's `rcon_password` changes:

```
19:24:50.222208 IP PubCompVM.27015 > localhost.48456: Flags [F.], seq 963797922, ack 2132329655, win 2048, options [nop,nop,TS val 2918891 ecr 2913896], length 0
    0x0000:  4500 0034 15e7 4000 4006 25db 7f00 0101  E..4..@.@.%.....
    0x0010:  7f00 0001 6987 bd48 3972 63a2 7f18 c4b7  ....i..H9rc.....
    0x0020:  8011 0800 ff28 0000 0101 080a 002c 89eb  .....(.......,..
    0x0030:  002c 7668                                .,vh
19:24:50.259394 IP localhost.48456 > PubCompVM.27015: Flags [.], ack 1, win 2050, options [nop,nop,TS val 2918901 ecr 2918891], length 0
    0x0000:  4500 0034 b2c7 4000 4006 88fa 7f00 0001  E..4..@.@.......
    0x0010:  7f00 0101 bd48 6987 7f18 c4b7 3972 63a3  .....Hi.....9rc.
    0x0020:  8010 0802 ff28 0000 0101 080a 002c 89f5  .....(.......,..
    0x0030:  002c 89eb      
```

Server sends FIN, client acknowledges. The socket technically is half-open, but _in the other direction_ from the way we thought it was. That's why Ruby was producing an `EPIPE` - it was still trying to write to the half of the connection that it had left open after the server closed its half. Node, on the other hand, doesn't allow half-open sockets unless you specifically tell it to, and automatically sends FIN back when it receives one from the server. Here's the result of an identical test using Node:

```
19:20:42.421160 IP PubCompVM.27015 > localhost.48455: Flags [F.], seq 1737808258, ack 2623120286, win 2048, options [nop,nop,TS val 2856941 ecr 2835379], length 0
    0x0000:  4500 0034 ef2a 4000 4006 4c97 7f00 0101  E..4.*@.@.L.....
    0x0010:  7f00 0001 6987 bd47 6794 d982 9c59 a39e  ....i..Gg....Y..
    0x0020:  8011 0800 ff28 0000 0101 080a 002b 97ed  .....(.......+..
    0x0030:  002b 43b3                                .+C.
19:20:42.421621 IP localhost.48455 > PubCompVM.27015: Flags [F.], seq 1, ack 1, win 2050, options [nop,nop,TS val 2856941 ecr 2856941], length 0
    0x0000:  4500 0034 9525 4000 4006 a69c 7f00 0001  E..4.%@.@.......
    0x0010:  7f00 0101 bd47 6987 9c59 a39e 6794 d983  .....Gi..Y..g...
    0x0020:  8011 0802 ff28 0000 0101 080a 002b 97ed  .....(.......+..
    0x0030:  002b 97ed                                .+..
19:20:42.421631 IP PubCompVM.27015 > localhost.48455: Flags [.], ack 2, win 2048, options [nop,nop,TS val 2856941 ecr 2856941], length 0
    0x0000:  4500 0034 ef2b 4000 4006 4c96 7f00 0101  E..4.+@.@.L.....
    0x0010:  7f00 0001 6987 bd47 6794 d983 9c59 a39f  ....i..Gg....Y..
    0x0020:  8010 0800 ff28 0000 0101 080a 002b 97ed  .....(.......+..
    0x0030:  002b 97ed                                .+..
```

Here, the client sends FIN as well, and the socket is fully closed on both ends. That's what the problem is... steam-condenser needs to close its half of the socket and then open a new socket after the password change. It's still trying to use the write-half of the old socket, which the server isn't bound to anymore, thus the `EPIPE`.

**EDIT:** SRCDS also does this right before a map change. I guess this is what we get for questioning the abilities of our overlords at Valve Software.

**Comment by koraktor** [comment]
Seems like I should've spend some more time testing this, with all three implementations and Wireshark. My time to work for in depth debugging is a bit limited right now, but I'll try to get this right soon.

**Comment by koraktor** [comment]
These patches should allow all three implementations to handle connection resets more gracefully. You will need to reauthenticate the connection.
I won't save RCON passwords for Source servers internally, though.

**Comment by koraktor** [comment]
@RJMiller Did you test this patch?

**Comment by smhxx** [comment]
Sorry about that, I've been a bit busy the past few days and haven't had a whole lot of time to check whether this patch has completely fixed the problem. It appears that the problem is at least partially fixed. When the client writes directly to `rcon_password`, the `rcon_exec` method returns empty string, prompting the user to re-authenticate. Everything looks good here:

```
irb(main):009:0> server.rcon_auth("password")
=> true
irb(main):010:0> server.rcon_exec("rcon_password password")
=> ""
irb(main):011:0> server.rcon_auth("password")
=> true
irb(main):012:0> server.rcon_exec("sv_cheats")
=> "\"sv_cheats\" = \"0\"\n notify replicated\n - Allow cheats on server"
```

However, I did still get exceptions in the following two situations:
- When the user sends a `changelevel` command, attempting to send a command after the server returns from the map change results in an `ECONNRESET`.
- If `rcon_password` is written to on the server console, or due to an automatic map change (such as one triggered by `mp_timelimit`), attempting to send a command results in an `IOError: Closed stream`. I think the problem here is that unless the client is _expecting_ a response from the server, it doesn't catch onto the fact that the socket got closed.

**Comment by koraktor** [comment]
Well, at least there's an improvement. :smirk:

Thanks for finding some more edge cases, I'll try to handle them too.

**Comment by koraktor** [comment]
This should at least fix the second problem you encountered. I couldn't reproduce the `ECONNRESET`, though.

**Comment by smhxx** [comment]
Hmm, I may have messed something up, I'll look into it.

**Comment by koraktor** [comment]
@RJMiller Any news on the connection reset?

**Comment by koraktor** [comment]
Please the the [latest comment on issue 176](https://github.com/koraktor/steam-condenser/issues/176#issuecomment-11464181): The connection reset on `changelevel` is most likely caused by setting the `rcon_password` inside `server.cfg`.

**Comment by koraktor** [comment]
This problem should be fixed by the patches above. Please try with the newest code from `master`.

**Comment by koraktor** [comment]
Closing this now. Please update this issue if you still have similar problems.

**Comment by koraktor** [comment]
@RJMiller Please note that there will be an additional change to this behavior released with version 1.3.1.

The specific commit is koraktor/steam-condenser-ruby@908dc6effdb7a4a0883f7fcad8526bb8936ab31b for the Ruby implementation. With that change a `changelevel` will now "fail" with a `RCONNoAuthError`. This seems like a step back first, but in total it is a more consistent and safer handling of connection resets. The best way to handle arbitrary RCON commands looks like follows:

``` ruby
if server.rcon_auth password
  begin
    server.rcon_exec command
  rescue RCONNOAuth
    if server.rcon_auth password
      # password is still correct, try again
    end
  end
else
  # password is wrong
end
```

**Comment by smhxx** [comment]
I apologize for my disappearance, I've been busy with other things and haven't had time to follow this as closely as I would have liked to. Your solution seems reasonable and is certainly a more elegant way of handling password changes than anything that would have been possible before. Good work.
